File: 885-exam-room/exam-room.java
import java.util.*;

class ExamRoom {
    private int n;
    private TreeSet<Integer> occupied;

    public ExamRoom(int n) {
        this.n = n;
        this.occupied = new TreeSet<>();
    }

    public int seat() {
        if (occupied.isEmpty()) {
            occupied.add(0);
            return 0;
        }

        int prev = -1;
        int maxDist = 0;
        int seat = 0;

        // Check seats between occupied seats
        for (int curr : occupied) {
            if (prev == -1) {
                // Check distance from 0 to the first occupied seat
                if (curr > maxDist) {
                    maxDist = curr;
                    seat = 0;
                }
            } else {
                // Check the middle of two occupied seats
                int dist = (curr - prev) / 2;
                if (dist > maxDist) {
                    maxDist = dist;
                    seat = prev + dist;
                }
            }
            prev = curr;
        }

        // Check the distance from the last occupied seat to the end
        if (n - 1 - occupied.last() > maxDist) {
            seat = n - 1;
        }

        occupied.add(seat);
        return seat;
    }

    public void leave(int p) {
        occupied.remove(p);
    }
}
/**
 * Your ExamRoom object will be instantiated and called as such:
 * ExamRoom obj = new ExamRoom(n);
 * int param_1 = obj.seat();
 * obj.leave(p);
 */

File: 885-exam-room/exam-room.py
from sortedcontainers import SortedList
class ExamRoom:
    """
      0 1 2 3 4 5 6 7 8 9
      ^   ^   ^         ^  
    """
    def __init__(self, n: int):
        self.n = n
        self.sorted_list = SortedList()

    def seat(self) -> int:

        if len(self.sorted_list) == 0:
            self.sorted_list.add(0)
            return 0

        mx_dist = 0
        res = 0
        prev = None
        # Seat in between seats
        for cur in self.sorted_list:
            if prev is None:
                prev = cur
                continue

            dist = cur - prev
            half = dist // 2

            if half > mx_dist:
                mx_dist = half
                res = prev + half
            prev = cur

        # Seat at seat 0?
        if self.sorted_list[0] >= mx_dist:
            mx_dist = self.sorted_list[0]
            res = 0
        
        # Seat at seat n - 1?
        if (self.n - 1 - self.sorted_list[len(self.sorted_list) - 1]) > mx_dist:
            res = self.n - 1
        
        self.sorted_list.add(res)
        return res

    def leave(self, p: int) -> None:
        self.sorted_list.remove(p)

# Your ExamRoom object will be instantiated and called as such:
# obj = ExamRoom(n)
# param_1 = obj.seat()
# obj.leave(p)